### PR TITLE
compatibility with fastn-stack/fastn/pull/1702

### PR DESCRIPTION
**Pull request by siddhantk232**

### CODE REVIEW

File: fastn/django/__init__.py [hunk 1]
```diff
@@ -110,13 +109,12 @@ def _add_user(self, request: RequestType):
...
+            username=fastn_user.get("username"),
             defaults={
                 "first_name": first_name,
                 "last_name": last_name,
-                "email": fastn_user.get("email") or "",
-                # email has Not Null constraint
-            }
+                "email": fastn_user.get("email"),
+            },
         )
```
Comment (amitu) [inline fastn/django/__init__.py]: Let's convert this to a hook, 

```py
import django.utils
from django.conf import settings

django.utils.module_loading(
    settings.get_attr(
        "FASTN_AUTH_CALLBACK", 
        "fastn.django.default_callback"
    )
)(request=request, fastn_user=fastn_user)
```

Where `fastn.django.default_callback` will be some code like this:

```py
def default_callback(request, fastn_user):
       user, _ = User.objects.get_or_create(
            username=fastn_user.get("login"),
            username=fastn_user.get("username"),
            defaults={
                "first_name": first_name,
                "last_name": last_name,
                "email": fastn_user.get("email") or "",
                # email has Not Null constraint
            }
                "email": fastn_user.get("email"),
            },
        )
```

Comment (siddhantk232) [reply fastn/django/__init__.py]: should I also add call to `django.contrib.auth.login` in `default_callback`?

Comment (amitu) [reply fastn/django/__init__.py]: Not sure. I think we will always need `auth.login()`.

Comment (amitu) [reply fastn/django/__init__.py]: What does it do?

Comment (siddhantk232) [reply fastn/django/__init__.py]: it creates a django session entry for the authenticated user.

>  Persist a user id and a backend in the request. This way a user doesn't
    have to reauthenticate on every request. Note that data set during
    the anonymous session is retained when the user logs in.

Comment (siddhantk232) [reply fastn/django/__init__.py]: if we choose to move `auth.login` inside the callback. Our middleware will just be responsible for decoding the fastn set cookie into a python dict and then the callback can decide to do whatever it likes with it which I think is a better way.

Comment (siddhantk232) [reply fastn/django/__init__.py]: we can also get rid of the utility function `utils.get_first_name_and_last_name`

Comment (amitu) [reply fastn/django/__init__.py]: Let's do this.

Comment (siddhantk232) [reply fastn/django/__init__.py]: done :)

File: fastn/django/__init__.py [hunk 2]
```diff
@@ -17,6 +18,10 @@
 CI = utils.AESCipher(SECRET_KEY)
 COOKIE_MAX_AGE = 365 * 24 * 60 * 60
 
+FASTN_AUTH_CALLBACK = getattr(
+    settings, "FASTN_AUTH_CALLBACK", "fastn.django.default_callback"
+)
```
Comment (amitu) [inline fastn/django/__init__.py:23]: You can do the `module_loading.import_string` resolution part here as well to only do it once, instead on per request.

Comment (siddhantk232) [reply fastn/django/__init__.py:23]: [done](https://github.com/fastn-stack/fastn_py/pull/1/commits/1e414934204bf04152314daf3dcc7e7c811a3ce6)

File: fastn/django/__init__.py [hunk 3]
```diff
@@ -31,32 +35,32 @@ def wrapper(request: RequestType):
...
 
 def redirect(location):
-    return django.http.JsonResponse({
-        "redirect": location,
-    })
+    return django.http.JsonResponse(
+        {
+            "redirect": location,
+        }
+    )
```
Comment (amitu) [inline fastn/django/__init__.py:52]: There is a redirect helper we can use.

File: fastn/django/__init__.py [hunk 4]
```diff
@@ -31,32 +35,32 @@ def wrapper(request: RequestType):
...
 
 def reload():
-    return django.http.JsonResponse({
-        "reload": True,
-    })
+    return django.http.JsonResponse(
+        {
+            "reload": True,
+        }
+    )
```
Comment (amitu) [inline fastn/django/__init__.py:60]: This should also use `reload()` helper.

**Pull request merged**